### PR TITLE
add AFL_TARGET_ENV to afl-fuzz

### PR DESCRIPTION
**Pull request by fuzzah**
This covers cases like specifying LD_LIBRARY_PATH only for the target. This also allows fuzzing targets that require AFL-specific env variables (e.g. AFL++ itself with AFL_BENCH_JUST_ONE).
I have fuzzed parsing function **extract_and_set_env** for ~2 hours on 6 cores and got hundreds of millions of execs and 100% coverage so it should be secure enough.
 
AFL_TARGET_ENV supports cases like
A=1
B='1 2 3'
C="x y z"
So quotes supported and should match. Outside quotes separator is ' ' (space).
Keys (vars) without values are not allowed and keys starting with numbers are not allowed either, so basically AFL_TARGET_ENV tries to follow UNIX rules. Keys with lowercase are allowed because some applications use them.
 
~~Length limit for value of AFL_TARGET_ENV is hardcoded to be 30000 characters.~~
upd: not hardcoded anymore

### CODE REVIEW

**Comment by fuzzah** [comment]
Also this fixes #774

File: src/afl-common.c [hunk 1]
```diff
@@ -706,6 +706,105 @@ char *get_afl_env(char *env) {
 
 }
 
+int extract_and_set_env(const char *env_str) {
+
+  size_t len = strnlen(env_str, 30000);
+  char * p = strndup(env_str, len);
```
Comment (vanhauser-thc) [inline src/afl-common.c]: please just use `u8 *p = ck_strdup(env_str);` instead, PFATAL can go too.

File: src/afl-common.c [hunk 2]
```diff
@@ -706,6 +706,105 @@ char *get_afl_env(char *env) {
...
+    num_pairs += 1;
+
+    setenv(key, val, 1);
+
+  }
+
+  if (num_pairs > 0) { ret_val = 1; }
+
+free_and_return:
+  free(p);
```
Comment (vanhauser-thc) [inline src/afl-common.c]: because of ck_strdup, this then needs to be ck_free

File: src/afl-common.c (hunk 1 above)
Comment (vanhauser-thc) [reply src/afl-common.c]: ah and `if (!env_str) { return 0; }` before that

File: src/afl-common.c [hunk 3]
```diff
@@ -706,6 +706,105 @@ char *get_afl_env(char *env) {
...
+
+  size_t len = strnlen(env_str, 30000);
+  char * p = strndup(env_str, len);
+
+  if (!p) { PFATAL("Unable to copy environment variables (OOM?)"); }
+
+  int ret_val = 0;  // return false by default
+
+  char *end = p + len;
+  *end = '\0';  // ensure null present
```
Comment (vanhauser-thc) [inline src/afl-common.c]: also please switch all `int` to `u32` and `char` to `u8`

**Comment by vanhauser-thc** [comment]
just a few nits, fixing the conflict and then it can be merged :)

**Comment by fuzzah** [comment]
> 
> 
> just a few nits, fixing the conflict and then it can be merged :)

Thank you for reviewing this.
Fixed & squashed.
Also fuzzed for 500M execs just in case :)

**Comment by fuzzah** [comment]
changed the last sneaky char to u8..
@vanhauser-thc i guess i shouldn't have squashed and force-pushed :( sorry for that

**Comment by vanhauser-thc** [comment]
LGTM, thank you!

**Pull request merged**